crates/ruff_python_parser/src/parser/statement.rs

```diff
@@ -2108,57 +2108,105 @@ impl<'src> Parser<'src> {
+        // match *x and y, z:
+        //     case _: ...
+        // match yield x:
+        //     case _: ...
         let subject = if self.at(TokenKind::Comma) {
             let tuple =
                 self.parse_tuple_expression(subject.expr, subject_start, Parenthesized::No, |p| {
-                    p.parse_named_expression_or_higher(AllowStarredExpression::No)
+                    p.parse_star_expression_or_higher(AllowNamedExpression::Yes)
                 });
 
             Expr::Tuple(tuple).into()
         } else {
+            if subject.is_unparenthesized_starred_expr() {
+                // test_err match_stmt_single_starred_subject
+                // match *foo:
+                //     case _: ...
+                self.add_error(ParseErrorType::StarredExpressionUsage, &subject);
+            }
             subject
         };
 
         self.expect(TokenKind::Colon);
 
-        self.eat(TokenKind::Newline);
+        // test_err match_stmt_no_newline_before_case
+        // match foo: case _: ...
+        self.expect(TokenKind::Newline);
+
+        // Use `eat` instead of `expect` for better error message.
         if !self.eat(TokenKind::Indent) {
-            let range = self.current_token_range();
+            // test_err match_stmt_expect_indented_block
+            // match foo:
+            // case _: ...
             self.add_error(
                 ParseErrorType::OtherError(
-                    "expected an indented block after `match` statement".to_string(),
+                    "Expected an indented block after `match` statement".to_string(),
                 ),
-                range,
+                self.current_token_range(),
             );
         }
 
-        let cases = self.parse_match_cases();
+        let cases = self.parse_match_case_blocks();
 
         self.eat(TokenKind::Dedent);
 
         ast::StmtMatch {
             subject: Box::new(subject.expr),
             cases,
-            range: self.node_range(start_offset),
+            range: self.node_range(start),
         }
     }
 
     /// Parses a list of match case blocks.
-    fn parse_match_cases(&mut self) -> Vec<ast::MatchCase> {
+    fn parse_match_case_blocks(&mut self) -> Vec<ast::MatchCase> {
+        let mut cases = vec![];
+
         if !self.at(TokenKind::Case) {
+            // test_err match_stmt_expected_case_block
+            // match x:
+            //     x = 1
+            // match x:
+            //     match y:
+            //         case _: ...
             self.add_error(
-                ParseErrorType::OtherError("expecting `case` block after `match`".to_string()),
+                ParseErrorType::OtherError("Expected `case` block".to_string()),
                 self.current_token_range(),
             );
+            return cases;
         }
 
-        let mut cases = vec![];
         let mut progress = ParserProgress::default();
 
         while self.at(TokenKind::Case) {
```

Comment:
Should this use the list parsing infrastructure to ensure that the statement recovery correctly recovers when seeing a `case` keyword:


```
match x:
    case "test":
        async
                
    case "b": # The statement recovery should not eat over the `case` keyword
```

Reply:
I don't have a specific example but something where the expression parsing goes into recovery. It then skips over all tokens that aren't the start of a statement or expression, which is why I would expect that it skips over the `case` keyword. 

Or does that not happen today because the `case` keyword is parsed as a `Name` rather than the `case` keyword because the soft keyword tokenizer gets confused?

Reply:
I think it should be ok because of the `Dedent` token? I'll need to think about this but I do understand what you're trying to say.

Reply:
> but I do understand what you're trying to say.

That's good haha

Reply:
Just to be clear, in case of any error in any of the case block, the parser should continue parsing the remaining case blocks as is, right?

Reply:
Yes, it shouldn't consume the `case` because it's a valid recovery point.